src/dune_rules/artifacts.ml

```diff
@@ -9,15 +9,19 @@ module Bin = struct
   type t =
     { context : Context.t
     ; (* Mapping from executable names to their actual path in the workspace.
-         The keys are the executable names without the .exe, even on Windows. *)
-      local_bins : Path.Build.t String.Map.t
+         The keys are the executable names without the .exe, even on Windows.
+         Enumerating binaries from install stanzas may involve expanding globs,
+         but the artifacts database is depended on by the logic which expands
+         globs. This field is a memo to break the cycle. *)
+      local_bins : Path.Build.t String.Map.t Memo.t
```

Comment:
Making this map isn't the fastest operation, and every bind will run the computation. Can we make this field lazy?

Reply:
I didn't realize that binding on a memo recomputes the value after the first time it's computed. Why is that necessary?

Reply:
Also how would making it lazy (I assume a `_ Memo.Lazy.t`, not a `_ Lazy.t`) help? Is there documentation somewhere on how `Memo.Lazy` works?

Reply:
> I didn't realize that binding on a memo recomputes the value after the first time it's computed. Why is that necessary?

Always saving the intermediate computation would be more expensive.

> Also how would making it lazy (I assume a _ Memo.Lazy.t, not a _ Lazy.t) help? Is there documentation somewhere on how Memo.Lazy works?

Yes, you need `Memo.Lazy.t`. As a first approximation, you can view it as a memoized function of type `unit -> 'a Memo.t`. The only advantage is that it's implemented a little more efficiently .

Reply:
Ah ok, so just to make sure I understand, is it the case that the only computations that are actually memoized by memo are the functions passed to `Memo.create`?

And is it correct that `Memo.lazy_` both creates a thunk returning a memo, and creates a cell to store the `'a` produced when the memo eventually computes a value? And the thunk closes over the cell, and after it is filled, subsequent calls to `force` return memos which don't require any additional computation. Is that right? I was confused by the name because I expected it to be analogous to the built-in `Lazy.t` type but it's more like a way to cache a `memo`.

Reply:
Yes, you got that right.